labs/optimizely-data-cli/oevents

```diff
@@ -0,0 +1,651 @@
+    ErrorEcho "An Optimizely API token may be passed via the --token command line argument, e.g."
+    ErrorEcho "  $ $CLI_NAME ls --token <token> --type decisions --date 2020-07-01"
+    ErrorEcho "...or set the OPTIMIZELY_API_TOKEN environment variable, e.g."
+    ErrorEcho "  $ export OPTIMIZELY_API_TOKEN=<token>"
+    ErrorEcho "  $ $CLI_NAME ls --type decisions --date 2020-07-01"
+    exit 1
+  fi
+
+  local auth_endpoint="https://api.optimizely.com/v2/export/credentials"
+
+  VerboseEcho "Requesting AWS credentials from $auth_endpoint"
+
+  local auth_api_response=$(curl -w "%{response_code}" -s -H "Authorization: Bearer $OPTIMIZELY_API_TOKEN" -X GET $auth_endpoint)
+
+  local auth_api_response_code="${auth_api_response:(-3)}"
+
+  if [ "$auth_api_response_code" != "200" ]; then
+    ErrorEcho "Error: received error response from Optimizely Authentication API: $auth_api_response"
+    exit 1
+  fi
+
+  local response_length="${#auth_api_response}"
+
+  auth_api_credential_str="${auth_api_response:0:((response_length-3))}"
+}
+
+# ExtractValueFromJSON
+# Extract a value from the passed json string.  The value to be extract is specified
+# by a JSON path, passed in the second parameter.
+function ExtractValueFromJSON()
+{
+  local json_str="$1"
+  local path="$2"
+
+  local val=$(echo "$json_str" | jq -r "$path")
+
+  if [[ -z "${val:-}" || "$val" == "null" ]]; then
+    ErrorEcho "Error: Unable to extract $path from $json_str"
+    exit 1
+  fi
+  
+  echo "$val"
+}
+
+# ParseAuthAPIResponse
+# Extract access key id and other important values from the Optimizely
+# Auth API request.  See:
+# https://docs.developers.optimizely.com/optimizely-data/docs/authentication-api
+function ParseAuthAPIResponse()
+{
+  local cred_str="${auth_api_credential_str:-{\}}"
+
+  # Extract AWS credentials from the JSON response.
+  # Export key id, access key, and session token variables so that they are visible to
+  # aws when invoked by this script
+  export AWS_ACCESS_KEY_ID=$(ExtractValueFromJSON "$cred_str" ".credentials.accessKeyId")
+  export AWS_SECRET_ACCESS_KEY=$(ExtractValueFromJSON "$cred_str" ".credentials.secretAccessKey")
+  export AWS_SESSION_TOKEN=$(ExtractValueFromJSON "$cred_str" ".credentials.sessionToken")
+  export AWS_SESSION_EXPIRATION=$(ExtractValueFromJSON "$cred_str" ".credentials.expiration")
+  S3_BASE_PATH=$(ExtractValueFromJSON "$cred_str" ".s3Path")
+
+  # Exit if we failed to extract any of the expected variables 
+  if [[ -z "$AWS_ACCESS_KEY_ID" || 
+        -z "$AWS_SECRET_ACCESS_KEY" ||
+        -z "$AWS_SESSION_TOKEN" ||
+        -z "$AWS_SESSION_EXPIRATION" || 
+        -z "$S3_BASE_PATH" ]]; then 
+    ErrorEcho "Error: Failed to extract one or more expected values from the Authentication API response"
+    exit 1
+  fi
+}
+
+# Authenticate
+# Retrieve AWS credentials from Optimizely's authentication API.  
+function Authenticate()
+{
+  MakeAuthAPIRequest
+  ParseAuthAPIResponse
+
+  authenticated=true
```

Comment:
It seems like this CLI makes use of global variables quite a bit. To be honest, we should avoid using global variables as much as possible. Global variables are useful for values that do not change often. However, it seems like a lot of functions alter the global state. Global variables should be used for things that are evaluated once (generally at start up) such as the detected architecture of the running system, or the verbosity of the output.

There are many unfortunate side effects to relying on dynamic global variables. Debugging can be very difficult for example and it can be challenging to reason about the current state of the program. I recommend that we refactor this approach to avoid depending on globals so that we have a better understanding of scope.

Reply:
Very helpful advice.  I refactored and removed the dependency on `AUTHENTICATED`.  I left the three global arrays (`DATE_RANGE`, `S3_RELATIVE_PATHS`, `S3_ABSOLUTE_PATHS`)  in order to maintain compatibility with bash 3.2.  I feel OK about this because the current control flow is linear: these variables are computed once, successively, each time the script is invoked.